### PR TITLE
Shared string interner

### PR DESCRIPTION
**Pull request by tpaschalis**
Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --sign-off flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.

### CODE REVIEW

File: model/intern/intern.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
+// Copyright 2019 The Prometheus Authors
```
Comment (rfratto) [inline model/intern/intern.go:1]: nit: update copyright :)

File: scrape/manager.go [hunk 2]
```diff
@@ -264,7 +265,7 @@ func (m *Manager) ApplyConfig(cfg *config.Config) error {
 			sp.stop()
 			delete(m.scrapePools, name)
 		} else if !reflect.DeepEqual(sp.config, cfg) {
-			err := sp.reload(cfg)
+			err := sp.reload(cfg, m.opts.UseSharedInterner)
```
Comment (rfratto) [inline scrape/manager.go:268]: nit: flags can never change between reloads, maybe store the option in the struct instead?

**Review by rfratto — APPROVED** [review]

File: cmd/prometheus/main.go [hunk 3]
```diff
@@ -176,6 +176,9 @@ func (c *flagConfig) setFeatureListOptions(logger log.Logger) error {
 			case "extra-scrape-metrics":
 				c.scrape.ExtraMetrics = true
 				level.Info(logger).Log("msg", "Experimental additional scrape metrics")
+			case "scrape-cache-interner":
```
Comment (rfratto) [inline cmd/prometheus/main.go:179]: Should this be given a more generic name? We might want to use the shared interner for the Prometheus Agent too.

Comment (tpaschalis) [reply cmd/prometheus/main.go]: Sure. I went with `share-string-interner`.

**Pull request closed without merging**